IBC.md

```diff
@@ -305,10 +308,101 @@ pub struct IbcPacket {
+[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
+which was
+[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
+to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
+future-proof interface for contracts, we will use an approach inspired by that
+work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
+that implements this.
+
+After quite some
+[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
+we identified a number of issues implementing this idea with the CosmWasm model:
+
+1. We only return `Ok(Response)` or `Err(String)`. There is no way to return an
+   arbitrary binary encoded message in the error variant.
+2. If we return `messages` to be dispatched, they may fail with an error, which
+   is returned directly to the x/wasm handler. Obviously an error from the bank
+   handler will not be in the proper IBC Acknowledgement format.
+3. We also realized that in some cases, the original `ibc_packet_receive` will
+   not even have the information needed to format the success response, so we
+   will need some delayed response-builder in this case also.
+
+eg. In the ICS27 spec, they define a packet to register a new account, the
+acknowledgement of that packet should contain the address of the new account on
+the receiving chain. If we implemented this in CosmWasm, the contract would have
+to return a submessage of `WasmMsg::Instantiate{}` and only get the proper
+address for the response in `reply`. Thus, we need to have some way to fully
+build the success acknowledgement after the fact as well.
+
+With this in mind, we designed the following workflow for the `x/wasm` runtime:
+
+1. Wrap the storage with a "local cache", so we don't immediately write out
+   changes
+2. Call `ibc_receive_packet`, get either `Err(String)` or
+   `Ok(Response + acknowledgement)`. If `Err`, jump to 4.
+3. Execute any `submessages` and `messages` returned with this response. If any
+   return `Err`, jump to 4.
+4. `receiveResult` is `Err(String)` if any of the above errored, or `Ok(Binary)`
+   with the original acknowledgement bytes if we have the success case.
+5. Call `ibc_encode_acknowledgement` with the `receiveResult`. The return value
+   is the acknowledgement packet to send to the calling chain. This should never
+   error.
+6. If `ReceiveResult` is `Ok`, commit this "storage cache". If it is `Err`,
+   revert those changes.
+7. Pass the acknowledgement bytes to the IBC router to return
 
-##### Standard Acknowledgement Format
+```rust
+#[entry_point]
+pub fn ibc_encode_acknowledgement(
+    // maybe we pass some data from reply via storage
+    deps: DepsMut,
+    env: Env,
+    // the original packet we got if we need some data from that
+    packet: IbcPacket,
+    // Either Ok(ibc_receive_packet acknowledgement) or Err(failure message)
+    receiveResult: Result<Binary, String>,
```

Comment:
I'd split those to allow accessing the "receive data" in the error case as well:

```suggestion
    // Data returned by `ibc_receive_packet`
    receive_data: Binary,
    // Either Ok(()) or Err(failure message)
    processing_result: Result<(), String>,
```

Reply:
Okay, so `ibc_receive_packet` may return `acknowledgement: Binary` or may error.

If it succeeded but a dispatched (sub-)message errored, then we have both an error (eg. from BankMsg) and an acknowledgement (from the original call).

We end up with:

```
receive_data: Option<Binary>,
processing_result: Result<(), String>,
```

Does that make sense?

Reply:
I would not call the ack encoder/creator entry point if `ibc_receive_packet` errors but instead fail the transaction execution (same discussions as with messages).

Then `receive_data` is always set (to whatever `ibc_receive_packet` returned).

Reply:
Continuing on discord. I have clearly failed to explain the IBC architecture design that we fit into.